### PR TITLE
fix: hide topology errors in line hit detection & optimizations

### PR DESCRIPTION
**Pull request by megawac**
This fixes an issue where jsts may throw `TopologyError` errors when building the intersections due to the snappable lines built having invalid geometry. We also do some minor performance work by 

1) Parsing the lines to jsts representations a single time, instead of each call through the loop
2) moving the `isSameLine` check prior to the `IntersectionOp` for some minor performance in the case we can avoid the work.

# Others

- [x] It's not a hack or at least an unauthorized hack :).
- [x] Everything in ticket description has been fixed.
- [x] The author of the MR has made its own review before assigning the reviewer (made this a month ago and has been thoroughly tested).

### CODE REVIEW

**Comment by oterral** [comment]
@megawac   thx for the PRs and sorry for the late answer. I review all of them and try to be responsive in the next month.

File: src/helper/getIntersectedLinesAndPoint.js [hunk 1]
```diff
@@ -15,14 +15,25 @@ const getIntersectedLinesAndPoint = (coordinate, lines, map, snapTolerance) => {
...
+    parsedLines.forEach(([lineB, parsedLineB]) => {
+      if (lineA === lineB || isSameLines(lineA, lineB, map)) {
+        return;
+      }
+
+      let intersections;
+      try {
+        intersections = OverlayOp.intersection(parsedLineA, parsedLineB);
+      } catch (e) {
+        return; // The OverlayOp will sometimes error with topology errors for certain lines
```
Comment (oterral) [inline src/helper/getIntersectedLinesAndPoint.js:32]: @megawac  Have you some more precise info about those topology error ? Could be nice to have an example.

**Review by oterral — CHANGES_REQUESTED** [review]
I've added a file for a unit test. Can you add one tests for the topolgy error?

File: src/helper/getIntersectedLinesAndPoint.js (hunk 1 above)
Comment (megawac) [reply src/helper/getIntersectedLinesAndPoint.js:32]: I seem to recall them occurring quite often when #267 was occurring... I'll see if I can track down if there was more context on internal tickets.

Comment (megawac) [reply src/helper/getIntersectedLinesAndPoint.js:32]: image

This is from the original ticket reported on our side. So in the case the case of the error, `parsedLineB` snap line appears to be invalid from the error below.

```
Uncaught TopologyException: found non-noded intersection between LINESTRING ( 9.952973188580247 2.35845900632583, 11.739231890263671 2.3680559374052663 ) and LINESTRING ( 9.952973188580248 2.35845900632583, 11.452360499376073 508.73944164968873 ) [ (9.952973188580248, 2.35845900632583, NaN) ]
    at g.checkValid (bundle.js)
    at E.checkValid (bundle.js)
    at E.checkValid (bundle.js)
    at A.computeOverlay (bundle.js)
    at A.getResultGeometry (bundle.js)
    at A.overlayOp (bundle.js)
    at E.getResultGeometry (bundle.js)
    at E.overlayOp (bundle.js)
    at A.intersection (bundle.js)
    at bundle.js
    at Array.forEach (<anonymous>)
    at bundle.js
    at Array.forEach (<anonymous>)
    at J (bundle.js)
    at ee.drawSnapLines (bundle.js)
    at ee.onMove (bundle.js)
    at a.handleEvent (bundle.js)
    at le.handleMapBrowserEvent (bundle.js)
    at u.dispatchEvent (bundle.js)
    at u.relayMoveEvent_ (bundle.js)
    at HTMLDivElement.d (bundle.js)
```

**Comment by oterral** [comment]
@megawac  can you fix the conflict? I will merge this even without test because it's usefull.

**Comment by megawac** [comment]
Done, the linting error appears to be an issue with a ci token

**Pull request merged**